core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```diff
@@ -339,156 +340,183 @@ class GroupMetadataManager(brokerId: Int,
-                    Errors.COORDINATOR_NOT_AVAILABLE
-
-                  case Errors.NOT_LEADER_OR_FOLLOWER
-                       | Errors.KAFKA_STORAGE_ERROR =>
-                    Errors.NOT_COORDINATOR
-
-                  case Errors.MESSAGE_TOO_LARGE
-                       | Errors.RECORD_LIST_TOO_LARGE
-                       | Errors.INVALID_FETCH_SIZE =>
-                    Errors.INVALID_COMMIT_OFFSET_SIZE
-
-                  case other => other
-                }
-              }
-            }
-
-            // compute the final error codes for the commit response
-            val commitStatus = offsetMetadata.map { case (topicIdPartition, offsetAndMetadata) =>
-              if (validateOffsetMetadataLength(offsetAndMetadata.metadata))
-                (topicIdPartition, responseError)
-              else
-                (topicIdPartition, Errors.OFFSET_METADATA_TOO_LARGE)
-            }
+      return
+    }
 
-            // finally trigger the callback logic passed from the API layer
-            responseCallback(commitStatus)
-          }
+    val magicOpt = getMagic(partitionFor(group.groupId))
+    if (magicOpt.isEmpty) {
+      val commitStatus = offsetMetadata.map { case (topicIdPartition, _) =>
+        (topicIdPartition, Errors.NOT_COORDINATOR)
+      }
+      responseCallback(commitStatus)
+      return
+    }
 
-          if (isTxnOffsetCommit) {
-            group.inLock {
-              addProducerGroup(producerId, group.groupId)
-              group.prepareTxnOffsetCommit(producerId, offsetMetadata)
-            }
-          } else {
-            group.inLock {
-              group.prepareOffsetCommit(offsetMetadata)
-            }
-          }
+    val isTxnOffsetCommit = producerId != RecordBatch.NO_PRODUCER_ID
+    val records = generateOffsetRecords(magicOpt.get, isTxnOffsetCommit, group.groupId, offsetTopicPartition, filteredOffsetMetadata, producerId, producerEpoch)
+    val putCacheCallback = createPutCacheCallback(isTxnOffsetCommit, group, consumerId, offsetMetadata, filteredOffsetMetadata, responseCallback, producerId, records)
 
-          appendForGroup(group, entries, requestLocal, putCacheCallback, transactionalId)
+    val verificationGuards = verificationGuard match {
+      case Some(guard) => Map(offsetTopicPartition -> guard)
+      case None => Map.empty[TopicPartition, VerificationGuard]
+    }
 
-        case None =>
-          val commitStatus = offsetMetadata.map { case (topicIdPartition, _) =>
-            (topicIdPartition, Errors.NOT_COORDINATOR)
-          }
-          responseCallback(commitStatus)
+    group.inLock {
+      if (isTxnOffsetCommit) {
+        addProducerGroup(producerId, group.groupId)
+        group.prepareTxnOffsetCommit(producerId, offsetMetadata)
+      } else {
+        group.prepareOffsetCommit(offsetMetadata)
       }
     }
+
+    appendForGroup(group, records, requestLocal, putCacheCallback, verificationGuards)
```

Comment:
What error do we expect if the guard check fails during write?

Reply:
We expect INVALID_TXN_STATE which is fatal. I think we previously discussed this and decided it was ok for old clients.

Reply:
We have logic in the group coordinator to convert the error code returned in TxnOffsetCommit. Do we want to add a case for INVALID_TXN_STATE?

Reply:
We considered this on the first PR but decided that the abortable/retriable errors were not specific enough.

From KIP-890 
> Return Abortable Error for TxnOffsetCommitRequests
Instead of  INVALID_TXN_STATE and INVALID_PID_MAPPING are we considered using UNKNOWN_MEMBER_ID which is abortable. However, this is not a clear message and is not guaranteed to be abortable on non-Java clients. Since we can't specify a message in the response, we thought it would be better to just send the actual (but fatal) errors.

Reply:
Ok. Mainly I was considering whether we should have an explicit case for this so that it is clearly intentional. What do you think?

Reply:
I can do that. I just thought that all the ones there were ones that were changed. There are also errors that are returned but not mapped. (ie, coordinator_not_available)

We can also include InvalidPidMapping if we do want to map errors.